### PR TITLE
bpo-1635741: sha256 heap types

### PR DESCRIPTION
**Pull request by koubaa**
https://bugs.python.org/issue1635741

### CODE REVIEW

**Review by vstinner — APPROVED** [review]
LGTM at the first iteration, good job :-) Let's see if the CI agrees with me.

**Comment by vstinner** [comment]
CI is unhappy:

```
Debug memory block at address p=0x7f408d1944a0: API 'm'
    0 bytes originally requested
    The 7 pad bytes at p-7 are FORBIDDENBYTE, as expected.
    The 8 pad bytes at tail=0x7f408d1944a0 are not all FORBIDDENBYTE (0xfd):
        at tail+0: 0x00 *** OUCH
        at tail+1: 0x00 *** OUCH
        at tail+2: 0x00 *** OUCH
        at tail+3: 0x00 *** OUCH
        at tail+4: 0x00 *** OUCH
        at tail+5: 0x00 *** OUCH
        at tail+6: 0x00 *** OUCH
        at tail+7: 0x00 *** OUCH

Enable tracemalloc to get the memory block allocation traceback

Fatal Python error: _PyMem_DebugRawFree: bad trailing pad byte
Python runtime state: finalizing (tstate=0x55e690e76d60)

Current thread 0x00007f408e7e0080 (most recent call first):
<no Python frame>
```

**Comment by vstinner** [comment]
Please try to build Python and run the test suite before pushing a change, run at least the modified test (test_hashlib in your case, if I followed correctly).

**Comment by koubaa** [comment]
@vstinner I did build locally and there were errors, so I pushed to see if it was windows-specific.  I haven't yet requested a review.

What is the proper etiquette for this in CPython, should I have added `[DRAFT]` or `[DONTREVIEW]` to the title?

**Comment by koubaa** [comment]
@vstinner @shihai1991 @corona10 please review

**Comment by vstinner** [comment]
> @vstinner I did build locally and there were errors, so I pushed to see if it was windows-specific.

If a test fails on Windows, a PR cannot be merged, since tests must pass on Linux and Windows.

In general, the advice is not publish a PR before it works as expected. It's rare that people propose incomplete/non-working PRs.

File: Modules/sha256module.c [hunk 1]
```diff
@@ -517,74 +533,27 @@ static PyMemberDef SHA_members[] = {
...
-    0,                  /*tp_richcompare*/
-    0,                  /*tp_weaklistoffset*/
-    0,                  /*tp_iter*/
-    0,                  /*tp_iternext*/
-    SHA_methods,        /* tp_methods */
-    SHA_members,        /* tp_members */
-    SHA_getseters,      /* tp_getset */
+static PyType_Spec sha224_type_spec = {
+    .name = "_sha256.sha224",
+    .basicsize =  sizeof(SHAobject),
```
Comment (vstinner) [inline Modules/sha256module.c]: nitpick:

```suggestion
    .basicsize = sizeof(SHAobject),
```

File: Modules/sha256module.c [hunk 2]
```diff
@@ -517,74 +533,27 @@ static PyMemberDef SHA_members[] = {
...
+    .flags = Py_TPFLAGS_DEFAULT,
+    .slots = sha256_types_slots
 };
 
+static PyType_Spec sha256_type_spec = {
+    .name = "_sha256.sha256",
+    .basicsize =  sizeof(SHAobject),
```
Comment (vstinner) [inline Modules/sha256module.c]: nitpick:

```suggestion
    .basicsize = sizeof(SHAobject),
```

File: Misc/NEWS.d/next/Core and Builtins/2020-09-07-09-45-47.bpo-1635741.QuDIut.rst [hunk 3]
```diff
@@ -0,0 +1,2 @@
+Convert the :mod:`_sha256` extension module types to heap types
+(:pep:`489`).
```
Comment (vstinner) [inline Misc/NEWS.d/next/Core and Builtins/2020-09-07-09-45-47.bpo-1635741.QuDIut.rst]: Hum, PEP 489 is unrelated to heap types: https://www.python.org/dev/peps/pep-0489/

And the _sha256 module already uses the multi-phase init API (PEP 489). Using static types with the multi-phase init API is unsafe.

```suggestion
Convert the :mod:`_sha256` extension module types to heap types.
```

**Review by vstinner — APPROVED** [review]
LGTM, but please address the minor comments.

**Pull request merged**